libp2p-identify/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+				                   .map(|(msg, _)| msg)
+				                   .map_err(|(err, _)| err)
+				                   .and_then(|msg| if let Some(msg) = msg {
+					Ok(Some(parse_proto_msg(msg)?))
+				} else {
+					Ok(None)
+				});
+
+				Box::new(future) as Box<_>
+			}
+
+			Endpoint::Listener => {
+				let info = self.information;
+
+				let listen_addrs = info.listen_addrs
+				                       .into_iter()
+				                       .map(|addr| addr.to_string().into_bytes())
+				                       .collect();
+
+				let mut message = structs_proto::Identify::new();
+				message.set_agentVersion(info.agent_version);
+				message.set_protocolVersion(info.protocol_version);
+				message.set_publicKey(info.public_key);
+				message.set_listenAddrs(listen_addrs);
+				message.set_observedAddr(info.observed_addr.to_string().into_bytes());
+				message.set_protocols(RepeatedField::from_vec(info.protocols));
+
+				let bytes = message.write_to_bytes().expect("we control the protobuf message");
+				let future = socket.send(bytes).map(|_| None);
+				Box::new(future) as Box<_>
+			}
+		}
+	}
+}
+
+// Turns a protobuf message into an `IdentifyInfo`. If something bad happens, turn it into
+// an `IoError`.
+fn parse_proto_msg(msg: BytesMut) -> Result<IdentifyInfo, IoError> {
+	match protobuf_parse_from_bytes::<structs_proto::Identify>(&msg) {
+		Ok(mut msg) => {
+			let listen_addrs = {
+				let mut addrs = Vec::new();
+				for addr in msg.take_listenAddrs().into_iter() {
+					addrs.push(bytes_to_multiaddr(addr)?);
+				}
+				addrs
+			};
+
+			let observed_addr = bytes_to_multiaddr(msg.take_observedAddr())?;
+
+			Ok(IdentifyInfo {
+				public_key: msg.take_publicKey(),
+				protocol_version: msg.take_protocolVersion(),
+				agent_version: msg.take_agentVersion(),
+				listen_addrs: listen_addrs,
+				observed_addr: observed_addr,
+				protocols: msg.take_protocols().into_vec(),
+			})
+		}
+
+		Err(err) => {
+			Err(IoError::new(IoErrorKind::InvalidData, err))
+		}
+	}
+}
+
+// Turn a `Vec<u8>` into a `Multiaddr`. If something bad happens, turn it into an `IoError`.
+fn bytes_to_multiaddr(bytes: Vec<u8>) -> Result<Multiaddr, IoError> {
+	let string = match String::from_utf8(bytes) {
+		Ok(b) => b,
+		Err(err) => return Err(IoError::new(IoErrorKind::InvalidData, err)),
+	};
+
+	match Multiaddr::new(&string) {
+		Ok(b) => Ok(b),
+		Err(err) => return Err(IoError::new(IoErrorKind::InvalidData, err)),
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	extern crate libp2p_tcp_transport;
+	extern crate tokio_core;
+
+	use self::libp2p_tcp_transport::TcpConfig;
+	use self::tokio_core::reactor::Core;
+	use IdentifyInfo;
+	use IdentifyProtocol;
+	use futures::{IntoFuture, Future, Stream};
+	use libp2p_swarm::Transport;
+	use libp2p_swarm::multiaddr::Multiaddr;
+
+	#[test]
+	fn basic() {
+		let mut core = Core::new().unwrap();
+		let tcp = TcpConfig::new(core.handle());
+		let with_proto = tcp.with_upgrade(IdentifyProtocol {
+			information: IdentifyInfo {
```

Comment:
Why do we have this struct with a single field instead of implementing `ProtocolUpgrade` for `IdentifyInfo` or using a newtype? Even a `fn new` would be nicer IMO.

Reply:
I admit that this is a bit arbitrary. I did it this way in case we'd need more configuration info in the future.

Reply:
Add `fn new` then